### DISCUSSION TITLE
Troubleshooting No bean of type [MyFactory] exists

### DISCUSSION THREAD

**Discussion by cancech** [micronaut-projects/micronaut-core#6733] [Q&A] [answered]
We have a framework that was originally creating using Spring/Dagger and I am now working toward migrating it to Micronaut. I've managed to convert the first part/library, with the only real "gotcha" being, that the bean availability management that was originally handled using Spring Configuration files or Dagger Modules (i.e.: beans for each mode defined in specific Config/Module), are now managed using Requires (i.e.: beans from an Active Config/Module are annotated Requires(env="Active")) at either the class or package level. From playing around in a sample app, this seems to work, and the updated library is happily compiling and I can see that the generated Micronaut meta classes are appearing as well.

By way of verifying that the switchover worked as expected I'm trying to create some MicronautTest classes to verify that the a given configuration (i.e.: Active) is actually loading, and the beans the library should be providing can be Injected into the MicronautTest class. Everything is compiling properly, however the issue occurs when I try to run the test, and I'm always running into an issue along the line of *No bean of type [MyFactory] exists*. It's not necessarily always the same Factory that it complains about (but as far as I can tell it's always complaining about a Factory of some kind or other). The factory it is complaining about always has what seem to be the appropriate generated Micronaut classes, such as:

* $MyFactory$Definition$Reference.class
* $MyFactory$Definition.class
* $MyFactory$ProvidesBean1$Definition$Reference.class
* $MyFactory$ProvidesBean1$Definition.class
* $MyFactory$ProvidesBean2$Definition$Reference.class
* $MyFactory$ProvidesBean2$Definition.class

As per the message from the exception, I've enabled trace logging 

`<logger name="io.micronaut.context.condition" level="TRACE"/>`

and while I'm seeing various logging messages for beans that it passes through, I'm never seeing the Factory that it complains about not being able to find.

We're using Gradle to pull in the Micronaut dependencies

```
plugins {
    id "io.micronaut.library"
}

project.ext {
    picocliVersion = '4.6.2'
    micronautVersion = '3.2.1'
    micronautTestVersion = '3.0.5'
}

dependencies {
    annotationProcessor "info.picocli:picocli-codegen:$picocliVersion"
    annotationProcessor "io.micronaut:micronaut-validation:$micronautVersion"
    annotationProcessor "io.micronaut:micronaut-inject-java:$micronautVersion"
    
    implementation "info.picocli:picocli:$picocliVersion"
    implementation "io.micronaut.picocli:micronaut-picocli:4.0.1"
    implementation "io.micronaut:micronaut-cli:1.3.7"
    implementation "io.micronaut:micronaut-runtime:$micronautVersion"
    implementation "io.micronaut:micronaut-inject:$micronautVersion"
    implementation "io.micronaut:micronaut-validation:$micronautVersion"
    implementation "io.micronaut:micronaut-core:$micronautVersion"
    implementation "io.micronaut:micronaut-websocket:$micronautVersion"
    
    testImplementation "io.micronaut.test:micronaut-test-core:$micronautTestVersion"
    testImplementation "io.micronaut.test:micronaut-test-junit5:$micronautTestVersion"
}

tasks.withType(JavaCompile) {
    options.encoding = 'UTF-8'
        options.compilerArgs = [
        '-Amicronaut.processing.incremental=true',
        '-Amicronaut.processing.annotations=com.lmcda.vista.*',
        '-parameters'
    ]
}
```

and the Factories are modeled based on examples I've seen in the documentation and based on what I've created in a little test app where I first played around with Micronaut.

```
@Factory
abstract class MyFactory {

    @Bean
    @Singleton
    @Named("Bean1")
    MyObject1 providesBean1(SomeParam param1, OtherParam param2) {
        return new MyObject1(param1, param2);
    }

    @Bean
    @Singleton
    @Named("Bean2")
    MyObject2 providesBean2(@Named("InternalObjBuilder")InternalObjBuilder builder, AnotherParam param3) {
        return builder.build(param3);
    }
}
```

At this point I've been fighting with this for a few days, and I'm very much as a loss as to what could be going on. Everything *looks* like it should work, but it's not and the error message as best as I can tell is just simply saying "it's not working" without any indication as to why. Any thoughts/ideas as to what could be going on? Any thing that I should look at or check?

Maybe I'm just not understanding the output? Below please find the full stacktrace I am receiving when I run the test.

```
    Message: No bean of type [com.lmcda.vista.core.di.active.ActiveSenderFactory] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
    Path Taken: Service.providesVistaMessagingSystemFactoryService([VistaMessagingSystemFactory domainVistaMessagingSystemFactory],VistaMessagingSystemFactory uiVistaMessagingSystemFactory) --> VistaMessagingSystemFactory.providesDomainMessagingSystemFactory(StateDispatcher stateRequestDispatcher,[ObjectSender stateChangeObjectSender])
        at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2364)
        at app//io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3282)
        at app//io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3268)
        at app//io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2821)
        at app//io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2783)
        at app//io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1639)
        at app//io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:1933)
        at app//io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1189)
        at app//com.lmcda.vista.core.di.active.$DomainMsgCommFactory$ProvidesDomainMessagingSystemFactory1$Definition.build(Unknown Source)
        at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2337)
        at app//io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3282)
        at app//io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3268)
        at app//io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2821)
        at app//io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2783)
        at app//io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1639)
        at app//io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:1933)
        at app//io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1189)
        at app//com.lmcda.vista.core.di.active.$DomainMsgCommFactory$ProvidesVistaMessagingSystemFactoryService3$Definition.build(Unknown Source)
        at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2337)
        at app//io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3686)
        at app//io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3592)
        at app//io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1357)
        at app//io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:916)
        at app//com.lmcda.vista.core.di.BehaviorBundleTest.setup(BehaviorBundleTest.java:58)

        Caused by:
        io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [com.lmcda.vista.core.di.active.ActiveSenderFactory] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
            at app//io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2796)
            at app//io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1576)
            at app//com.lmcda.vista.core.di.active.$ActiveSenderFactory$ProvidesStateChangeSender0$Definition.build(Unknown Source)
            at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2337)
            ... 23 more
```

**Comment by graemerocher** [comment, marked as answer]
As described in https://docs.micronaut.io/latest/guide/#conditionalBeans try add logging for conditional beans and see what is failing that is causing the bean to be disabled:

```xml
<logger name="io.micronaut.context.condition" level="DEBUG"/>
```

**Reply by graemerocher** [reply]
I noticed as well your factory is `abstract` which could be the problem

**Reply by cancech** [reply]
So, yep... it was the `abstract`. Removed that and suddenly everything started working the way I had expected.

Thanks!

**Comment by rkrishnas81** [comment]
grails 6 jboss 7.4
getting failed: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [io.micronaut.context.event.ApplicationEventPublisher<io.micronaut.context.event.StartupEvent>] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 
my build.gradle. please help me
buildscript {
    repositories {
        maven { url "https://repo.grails.org/grails/core" }
        mavenCentral()
    }
    dependencies { // Not Published to Gradle Plugin Portal
        classpath "org.grails:grails-gradle-plugin:6.2.4"
        classpath "org.grails.plugins:hibernate5:8.1.0"
    }
}

plugins {
    id "groovy"
    id "com.github.erdi.webdriver-binaries" version "3.2"
    id "war"
    id "idea"
    id "com.bertramlabs.asset-pipeline" version "4.5.1"
    id "application"
    id "eclipse"
}

// Not Published to Gradle Plugin Portal
apply plugin: 'war'
apply plugin: 'org.grails.grails-web'

bootWar {
    enabled = true
}
jar {
    enabled = false
}
group = "com.example"

repositories {
    mavenCentral()
    maven { url "https://repo.grails.org/grails/core/" }
}

configurations {
    all {
        resolutionStrategy.eachDependency { DependencyResolveDetails details->
            if (details.requested.group == 'org.seleniumhq.selenium') {
                details.useVersion('4.19.1')
            }
        }
    }
}

dependencies {

    profile "org.grails.profiles:web"

        implementation('org.springframework.boot:spring-boot-starter-web') {
            exclude group: 'org.springframework.boot', module: 'spring-boot-starter-tomcat'
        }
    implementation "org.grails:grails-core"
    implementation "org.grails:grails-logging"
    implementation "org.grails:grails-plugin-databinding"
    implementation "org.grails:grails-plugin-i18n"
    implementation "org.grails:grails-plugin-interceptors"
    implementation "org.grails:grails-plugin-rest"
    implementation "org.grails:grails-plugin-services"
    implementation "org.grails:grails-plugin-url-mappings"
    implementation "org.grails:grails-web-boot"
    implementation "org.grails.plugins:gsp"
    implementation "org.grails.plugins:hibernate5"
    implementation "org.grails.plugins:scaffolding"
    implementation "org.springframework.boot:spring-boot-autoconfigure"
    implementation "org.springframework.boot:spring-boot-starter"
    implementation "org.springframework.boot:spring-boot-starter-actuator"
    implementation "org.springframework.boot:spring-boot-starter-logging"
    //implementation "org.springframework.boot:spring-boot-starter-tomcat"
    providedCompile 'org.springframework.boot:spring-boot-starter-tomcat'
    implementation "org.springframework.boot:spring-boot-starter-validation"
    compileOnly "io.micronaut:micronaut-inject-groovy"
    console "org.grails:grails-console"
    runtimeOnly "com.bertramlabs.plugins:asset-pipeline-grails:4.5.1"
    runtimeOnly "com.h2database:h2"
    runtimeOnly "org.apache.tomcat:tomcat-jdbc"
    runtimeOnly "org.fusesource.jansi:jansi:1.18"
    testImplementation "io.micronaut:micronaut-inject-groovy"
    testImplementation "org.grails:grails-gorm-testing-support"
    testImplementation "org.grails:grails-web-testing-support"
    testImplementation "org.grails.plugins:geb"
    testImplementation "org.seleniumhq.selenium:selenium-api:4.19.1"
    testImplementation "org.seleniumhq.selenium:selenium-remote-driver:4.19.1"
    testImplementation "org.seleniumhq.selenium:selenium-support:4.19.1"
    testImplementation "org.spockframework:spock-core"
    testRuntimeOnly "org.seleniumhq.selenium:selenium-chrome-driver:4.19.1"
    testRuntimeOnly "org.seleniumhq.selenium:selenium-firefox-driver:4.19.1"
    testRuntimeOnly "org.seleniumhq.selenium:selenium-safari-driver:4.19.1"
}

application {
    mainClass = "com.example.Application"
}

java {
    sourceCompatibility = JavaVersion.toVersion("11")
}

tasks.withType(Test) {
    useJUnitPlatform()
    systemProperty "geb.env", System.getProperty('geb.env')
    systemProperty "geb.build.reportsDir", reporting.file("geb/integrationTest")
    systemProperty 'webdriver.chrome.driver', "${System.getenv('CHROMEWEBDRIVER')}/chromedriver"
    systemProperty 'webdriver.gecko.driver', "${System.getenv('GECKOWEBDRIVER')}/geckodriver"
}
webdriverBinaries {
    chromedriver '122.0.6260.0'
    geckodriver '0.33.0'
    edgedriver '110.0.1587.57'
}
assets {
    minifyJs = true
    minifyCss = true
    skipNonDigests = true
}

**Comment by rkrishnas81** [comment]
using jdk17
